modules/core/src/test/java/org/apache/ignite/util/FailureSimulatingTcpDiscoverySpi.java

```diff
@@ -0,0 +1,404 @@
+
+    /** {@inheritDoc} */
+    @Override protected Socket createSocket0() throws IOException {
+        return new SocketWrap(super.createSocket0());
+    }
+
+    /**
+     * Enables simulation of network timeout.
+     *
+     * @param direction If negative, enables timeout simulation for incomming traffic. If positive, enables timeout
+     *                  simulation for outgoing traffic. Set 0 to simlate failure for both traffics.
+     * @param delay     Milliseconds of awaiting before raising {@code SocketTimeoutException}.
+     * @see SocketWrap#simulateTimeout(Socket, int)
+     */
+    public void enableNetworkTimeoutSimulation(int direction, int delay) {
+        simulatedTimeout = new IgnitePair<>(direction, delay);
+    }
+
+    /**
+     * Simulates network timeout if enabled, raises {@code SocketTimeoutException}.
+     *
+     * @param sock         The socket to simulate failure at.
+     * @param forceTimeout If positive of 0, overrides the delay preset in {@link #enableNetworkTimeoutSimulation(int,
+     *                     int)}.
+     * @see #enableNetworkTimeoutSimulation(int, int)
+     */
+    private void simulateTimeout(Socket sock, int forceTimeout) throws SocketTimeoutException {
+        IgnitePair<Integer> simulatedTimeout = this.simulatedTimeout;
+
+        if (simulatedTimeout == null)
+            return;
+
+        boolean isClientSock = sock.getLocalPort() < locPort || sock.getLocalPort() > locPort + locPortRange;
+
+        if (isClientSock && simulatedTimeout.get1() < 0 || !isClientSock && simulatedTimeout.get1() > 0)
+            return;
+
+        int timeout = forceTimeout >= 0 ? forceTimeout : simulatedTimeout.get2();
+
+        try {
+            Thread.sleep(timeout);
+        }
+        catch (InterruptedException ignored) {
+            // No-op.
+        }
+
+        throw new SocketTimeoutException("Simulated failure after delay: " + timeout + "ms.");
+    }
+
+    /**
+     * @see #simulateTimeout(Socket, int)
+     */
+    private void simulateTimeout(Socket sock) throws SocketTimeoutException {
+        simulateTimeout(sock, -1);
+    }
+
+    /**
+     * Network failure simulator.
+     */
+    private class SocketWrap extends Socket {
+        /** The real socket to simulate failure of. */
+        private final Socket delegate;
+
+        /**
+         * Constructor.
+         *
+         * @param sock The real socket to simulate failure of.
+         */
+        private SocketWrap(Socket sock) {
+            delegate = sock;
+        }
+
+        /** {@inheritDoc} */
+        @Override public OutputStream getOutputStream() throws IOException {
+            OutputStream src = delegate.getOutputStream();
+
+            return new OutputStream() {
+                /** {@inheritDoc} */
+                @Override public void write(@NotNull byte[] b) throws IOException {
+                    simulateTimeout(delegate);
+
+                    src.write(b);
```

Comment:
Real TcpDiscoverySPI woul require simulate-failure test-only methods. Was declared by previous reviwe. That's why I made separate SPI based only on socket. Look OK to simulate failure directly by socket, not by higher logic. Exceptions are thrown from socket/streams in reality. We simulate exceptions from socket/streams.

Reply:
Still disagree with the approach.
In case we emulate filtering, we should just do this.

Just set up forward filtering. Will writing to the socket cause an exception in this case?